File: composition/JavaScript/pipeline/src/domainModel.js
import { Result } from "./result";

const INVALID_NAME = "invalid name";
const INVALID_EMAIL_ADDRESS = "invalid email address";
const VALID_NAME = "John";
const VALID_EMAIL_ADDRESS = "jk@test.com";

function validateName(data) {
    if (data === null || !data.name || data.name === null)
        return new Result(INVALID_NAME, data);
    return new Result("", data);
}

function validateEmail(data) {
    if (data === null || !data.emailAddress || data.emailAddress === null)
        return new Result(INVALID_EMAIL_ADDRESS, data);
    return new Result("", data);
}

function updateEmailAddress(data) {
    const newInstanceWithDifferentEmailAddress = Object.assign({}, data, { emailAddress: VALID_EMAIL_ADDRESS });
    return new Result("", newInstanceWithDifferentEmailAddress);
}

File: composition/python/domain_model.py
from common.email_addresses import is_valid_email_address
from common.pipeline import Result

INVALID_NAME_MESSAGE = 'Invalid name'
EMAIL_ADDRESS_FORMAT_IS_INCORRECT_MESSAGE = 'Email address format is incorrect'


class Person:
    def __init__(self, name=None, email_address=None):
        self.name = name
        self.email_address = email_address


def validate_name(person):
    if not person.name:
        return Result(INVALID_NAME_MESSAGE, person)
    return Result(data=person)


def validate_email_address(person):
    if not is_valid_email_address(person.email_address):
        return Result(EMAIL_ADDRESS_FORMAT_IS_INCORRECT_MESSAGE, person)
    return Result(data=person)


def update_name(new_name):
    def partial(person):
        return Result(data=Person(new_name, person.email_address))
    return partial
